### ISSUE TITLE
Sample weights (weighted loss) for classification

### ISSUE THREAD

**Issue by nzhiltsov** [facebookresearch/fastText#278]
I'm wondering how one can add sample weights to increase importance of some training examples in 'fasttext supervised' mode? See analogy e.g. 'sample_weight' array parameter while doing model.fit in Keras or compute accuracy_score in sklearn.

**Comment by dcsan** [comment]
did you ever figure this out?

I'm also interested if there's a way to add some other type of metadata as hints for training, and have this be more important. for example a topic. if sentences are within a topic, this should have a more weighted matching factor...

**Comment by pommedeterresautee** [comment]
For adding a topic just add a word with the name of the topic. It will be associated with a vector that will participate to the construction of the model.

**Comment by dcsan** [comment]
@pommedeterresautee that will add the topic, but i wanted to have the metadata have an extreme weighting on the classification. almost to the point that absence of that metadata would remove it from that category. I guess i could do that with some prefiltering... but wanted to keep a single pipeline.

**Comment by matanster** [comment]
I've spent some time on implementing this today, so far something terribly wrong in my loss function. Here's my repo where training data is treated as weighted ([https://github.com/copula-ai/fastText/blob/master/weighted-input-data](toy training data here)) and the weights trickle all the way down to [where the loss and gradients are computed/applied](https://github.com/copula-ai/fastText/blob/72ae17bc27f2cfb83ab22eeb93d4d55553c698a2/src/model.cc#L124-L143).

This fork breaks the tight coupling of the embedding and classification modes, introducing specific functions for handling classification. In a certain cheeky vein, it seems the original repo made it a point to conflate the two, whereas except for sharing some matrix util functions, there's very little to topically relate them other than the concurrent HOGWILD-inspired multi-threaded implementation. You can use the classification algorithm on top any word embeddings you have regardless of the fasttext-specific embedding model; so I think this actually improves modularity, but doubt anyone would merge a PR with this deviation.

**Comment by Meekohi** [comment]
I'm curious about this as well. Our dataset has 95% 'positive' (A) examples and only 5% 'negative' (B) examples -- the model often converges to just "always choose A".

**Comment by cramdoulfa** [comment]
Also curious if there is a clean solution to that problem implemented

One workaround it is to duplicate some training samples proportionally to the desired weights, but it only works for integer weights and it's not great.

**Comment by Whadup** [comment]
https://github.com/Whadup/weightedFastText I tried it here, it's not up-to-date with the current fasttext version though

**Comment by carlos-rafael** [comment]
Hello, everyone. Is here any other update on this subect (Adding weights to specific subjects)?
